### DISCUSSION TITLE
About initial build and unit test

### DISCUSSION THREAD

**Discussion by ahmetakman** [lava-nc/lava#189] [Q&A]
During the installation procedure, for the build command `pyb -E unit` there happened a bunch of Assertation and Type errors while executing the unit tests from Python modules. 
Then the info occurs as Executed 180 unit tests and All unit tests passed. According to the error this happens in run_spk. Then build system gives same errors and Infos 3 times. 

Just after that the system warn about some test coverages are below 70%.
[INFO]  Overall lava-nc coverage is 72%
[INFO]  Overall lava-nc branch coverage is 66%
[INFO]  Overall lava-nc partial branch coverage is 91%

Finally, there occurs the error " BUILD FAILED - Error while executing setup command ['bdist_dumb']. " and build finishes.

I am using Python 3.9.6 
Windows 10. 
Intel I7 9750H

What could have caused this build error sequence what can be the solution?

Thanks,

**Comment by mgkwill** [comment]
Hi @ahmetakman thanks for the question.

Can you try to run the unit tests directly `python -m unittest` and report back the results? 

I have a feeling the pybuilder virtual environment is not working correctly on your system. Sometimes this can be caused by path limitations on Windows. We need to update all of our docs to running the unit tests directly rather than through pyb.

**Reply by ahmetakman** [reply]
I have removed all the setup and tried a clean one. But, I noticed that I had not done the build step successfully. Unfortunately, for the `pip install e . ` command, I am getting the same error   BUILD FAILED - Error while executing setup command ['bdist_dumb']. . Then even though running unit test on failed build unnecessary, I have tried to run unit test, and it has failed all of them naturally. 

BTW I am using the home directory, I do not know whether you mean that by path limitation. 

Sorry for the late answer, but the problem still continues.

**Reply by ahmetakman** [reply]
I am on way to solve the problem, the virtual environments does not get activated with the command given in the instruction in Windows Powershell, instead I have used `python3_venv\Scripts\activate` and it (seems) succesfully activated the environment. But, the build fails again with the same error while it was preparing meradata for build wheel.

**Comment by PaoloGCD** [comment]
Hi, I am having the same error when using pyb and directly by running python -m unittest. The result when running the latter is the following:

======================================================================
ERROR: test_sigma_decoding_float (tests.lava.proc.sdn.test_models.TestSigmaModels)
Test sigma decoding with cumulative sum.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/pcachi/lava/tests/lava/proc/sdn/test_models.py", line 67, in test_sigma_decoding_float
    input, output = self.run_test(
  File "/home/pcachi/lava/tests/lava/proc/sdn/test_models.py", line 42, in run_test
    sigma.run(condition=run_condition, run_cfg=run_config)
  File "/home/pcachi/lava/lava/magma/core/process/process.py", line 422, in run
    self._runtime.initialize()
  File "/home/pcachi/lava/lava/magma/runtime/runtime.py", line 136, in initialize
    self._build_message_infrastructure()
  File "/home/pcachi/lava/lava/magma/runtime/runtime.py", line 162, in _build_message_infrastructure
    self._messaging_infrastructure.start()
  File "/home/pcachi/lava/lava/magma/runtime/message_infrastructure/multiprocessing.py", line 72, in start
    self._smm.start()
  File "/home/pcachi/anaconda3/envs/lava/lib/python3.10/multiprocessing/managers.py", line 552, in start
    reader, writer = connection.Pipe(duplex=False)
  File "/home/pcachi/anaconda3/envs/lava/lib/python3.10/multiprocessing/connection.py", line 532, in Pipe
    fd1, fd2 = os.pipe()
OSError: [Errno 24] Too many open files

----------------------------------------------------------------------
Ran 205 tests in 15.245s

FAILED (errors=42, skipped=6)

I am using ubuntu 18.04.

**Reply by mgkwill** [reply]
> Hi, I am having the same error when using pyb and directly by running python -m unittest. The result when running the latter is the following:
> 
> ## ======================================================================
> ERROR: test_sigma_decoding_float (tests.lava.proc.sdn.test_models.TestSigmaModels)
> Test sigma decoding with cumulative sum.
> Traceback (most recent call last): File "/home/pcachi/lava/tests/lava/proc/sdn/test_models.py", line 67, in test_sigma_decoding_float input, output = self.run_test( File "/home/pcachi/lava/tests/lava/proc/sdn/test_models.py", line 42, in run_test sigma.run(condition=run_condition, run_cfg=run_config) File "/home/pcachi/lava/lava/magma/core/process/process.py", line 422, in run self._runtime.initialize() File "/home/pcachi/lava/lava/magma/runtime/runtime.py", line 136, in initialize self._build_message_infrastructure() File "/home/pcachi/lava/lava/magma/runtime/runtime.py", line 162, in _build_message_infrastructure self._messaging_infrastructure.start() File "/home/pcachi/lava/lava/magma/runtime/message_infrastructure/multiprocessing.py", line 72, in start self._smm.start() File "/home/pcachi/anaconda3/envs/lava/lib/python3.10/multiprocessing/managers.py", line 552, in start reader, writer = connection.Pipe(duplex=False) File "/home/pcachi/anaconda3/envs/lava/lib/python3.10/multiprocessing/connection.py", line 532, in Pipe fd1, fd2 = os.pipe() OSError: [Errno 24] Too many open files
> 
> Ran 205 tests in 15.245s
> 
> FAILED (errors=42, skipped=6)
> 
> I am using ubuntu 18.04.

Hi @PaoloGCD. Thanks for your question. It sounds like your issue is different than the OP. Have you tried the install solution in the FAQ https://github.com/lava-nc/lava/wiki/Frequently-Asked-Questions-(FAQ)#install ?

**Reply by PaoloGCD** [reply]
Thank you, yes my problem was the max number of files.
